### ISSUE TITLE
Navigation sample

### ISSUE THREAD

**Issue by hansmbakker** [windows-toolkit/MVVM-Samples#21]
It would be great to have a sample focusing on navigation between pages, showing how to pass data between them, how to create the ViewModels using IoC and wiring up the ViewModels with the page views the user is navigating to.

**Comment by Sergio0694** [comment]
That is a great suggestion, we did plan to also add some docs to show the process of creating and using a navigation service, at least as an example of how to go about writing platform specific services. This one in particular is used pretty often, so it would make sense to have a sample for it I think (at least with the basic concepts).

cc. @michael-hawker should we add an actual working sample in the app or just a docs page for this?

**Comment by michael-hawker** [comment]
@Sergio0694 I think both would be good since it's a common ask. 🙂

**Comment by ThomasDallmeier** [comment]
Hi anything new with a navigation service ?

**Comment by jader1313** [comment]
As long as the "official" example does not come out, could anyone indicate an example using WTC.MVVM? I am suffering a lot trying to implement this.

**Comment by Sergio0694** [comment]
I'll put together a minimal sample and post it here as a reference, at least for now.
Will update in a bit 😄

**Comment by Gerardo-Sista** [comment]
Yes, a small sample app shgowing navigation should be really appreciated.!

**Comment by jasonxz** [comment]
A sample including url-based navigation would be really nice.

**Comment by Javier118** [comment]
I wanted to ask how to do Navigate(typeof(Page2));

**Comment by oscarjaergren** [comment]
Yeah a sample would be good

**Comment by meenanarendra** [comment]
Please provide a WPF sample application with Navigation.

**Comment by pradeepkumar-devaraj** [comment]
I currently work on migrating our WPF application from MVVM Light to MVVM Toolkit. I am stuck at implementing the navigation functionality. It would be great if a WPF sample application is provided. Thanks.

**Comment by michael-hawker** [comment]
@Sergio0694 I know you've been sprucing up the sample app, is this on your list of things to add there?

**Comment by lloydjatkinson** [comment]
It would be good if MVVM Toolkit had a navigation mechanism already built in...

**Comment by hassan-gasemi** [comment]
@Sergio0694 Hi. Is there a best practice for setting rootFrame? In your snippet, rootFrame is set in App.xaml but in Wpf our frame usually  lay in MainWindow.

**Comment by curia-damiano** [comment]
+1 interested in having this feature in the dotnet Community Toolkit, thanks!

**Comment by snow-jallen** [comment]
Any update on navigation in CommunityToolkit?

**Comment by Sergio0694** [comment]
There are no plans to add built-in APIs related to navigation at this time. We'd be willing to accept a community contribution adding a navigation sample, provided it meets all requirements for code quality, documentation, etc. 🙂

**Comment by lloydjatkinson** [comment]
What is the reasoning behind not providing this?

**Comment by michael-hawker** [comment]
@lloydjatkinson we don't want to add specific APIs to the base library as Navigation tends to be platform specific for its implementation, and the library is platform-agnostic.

However, we welcome a sample that demonstrates how to do this pattern within the framework of our sample app. We just haven't had time to invest in this space above, though @Sergio0694 provided an example above from one of his own projects.

If someone wants to contribute a PR for this scenario based on that or some other implementation to the sample app repo, please feel free, as Sergio calls out above as well.

**Comment by mrlacey** [comment]
In addition to navigation often being platform specific, there is no one right or even best way to do this.
For example, there are solid arguments for and against navigation being between parts of the UI or between ViewModels. Similarly, there are arguments for and against coupling views and view models on a one-to-one basis. There is no one right way to do this, and the approach taken is often driven as much by personal preference as the application's needs. 
If only a single navigation example were provided, it might easily be interpreted as being the way that must or "should" be used. Without a solid explanation of where, when, and why different approaches may be more or less appropriate, any single sample technique could encourage people to use something that is a poor choice for their application.

**Comment by mstasak** [comment]
One example can be found if you create a Template Studio project (Windows App SDK) and allow it to generate a navigation service.  It is based on ViewModels, not Views/Pages.  Sadly, at present it doesn't really show you how to cancel a navigation or receive an event before the navigation occurs (so as to explicitly update bindings and/or commit changes to a database, or prompt the user to save/discard/cancel).  Anyway, it is a study-able example, and not too complex.

**Comment by mstasak** [comment]
Yes, even if there's no one superior approach for all situations, it would help to have some starting point, and perhaps a discussion of the alternative approaches and their benefits and pitfalls.  SPA with a redirectable content frame is a fairly trivial approach, but you still need to decide between a forward-back stack navigation or a freely addressable network of content targets.  You still need to decide if a content target will be abandoned/lost or persisted or reconstructed, should the user return to that target.  Add the possibilities of multiple modeless windows or modal popup windows, or even  multiple instances of the same view over different data (say a file or a database row) within your nav hierarchy, and the right approach could get very muddy.  It should still be possible to enumerate the common approaches, provide descriptions and examples, and perhaps craft a hybrid service which can reconfigure to cover all or most of these.  Not easy, but no reason to wring one's hands and say it's simply impossible or a bad idea.

**Comment by CodingOctocat** [comment]
@Sergio0694 If the Frame is defined in App.xaml.cs, how is it rendered in xmal? If I have a TabControl where each TabItem is a Frame, how do I implement it?

**Comment by IlyaNavodkin** [comment]
> I'll put together a minimal sample and post it here as a reference, at least for now. Will update in a bit 😄
> 
> **EDIT:** alright, so here's a few examples. One of the things I mentioned before was that there's really a number of ways to go about doing this, and people might want to tweak the implementation depending on their specific needs. For example, some developers might not be interested in having a platform-agnostic backend, so they might want to add framework-specific types directly on the navigation service (for instance, by passing the page type directly). Others might want to abstract this instead, so in that case you might want to have a service implementation that internally stores a dictionary of sorts to map viewmodel types to page types, and then perform the navigation directly. Some might want a more "pattern oriented" approach, for instance using some kind of standard naming practice for viewmodels, etc.
> 
> Here's the simples possible approach for this, with a framework-specific solution (wouldn't recommend this):
> 
> ```cs
> // Navigation interface
> public interface INavigationService
> {
>     bool CanGoBack { get; }
>     void GoBack();
>     void Navigate<T>(object args = null);
> }
> 
> // UWP navigation service
> public class NavigationService : INavigationService
> {
>     private readonly Frame frame;
> 
>     public NavigationService(Frame frame)
>     {
>         this.frame = frame;
>     }
> 
>     public bool CanGoBack => this.frame.CanGoBack;
> 
>     public void GoBack() => this.frame.GoBack();
> 
>     public void Navigate<T>(object args)
>     {
>         this.frame.Navigate(typeof(T), args);
>     }
> }
> ```
> 
> And then you'd use it something like this, for instance (eg. in a single-frame application):
> 
> ```cs
> // ...This is from App.xaml.cs, as usual
> rootFrame = new Frame();
> 
> // Configure your services somewhere after that point, and add the navigation service too
> Ioc.Default.ConfigureServices(
>     new ServiceCollection()
>     .AddSingleton<INavigationService>(new NavigationService(rootFrame))
>     .BuildServiceProvider());
> 
> // ...Then whenever you need to navigate, you can do:
> Ioc.Default.GetRequiredService<INavigationService>().Navigate<SomePage>();
> 
> // ...Or, using dependency injection in your viewmodels
> this.navigationService.Navigate<SomePage>();
> ```
> 
> A better approach could be, for instance, to have the navigation service store a mapping of viewmodel types:
> 
> ```cs
> public class NavigationService : INavigationService
> {
>     private readonly Dictionary<Type, Type> viewMapping = new()
>     {
>         [typeof(MainPageViewModel)] = typeof(MainPage),
>         // Other viewmodel types...
>     };
> 
>     private readonly Frame frame;
> 
>     public NavigationService(Frame frame)
>     {
>         this.frame = frame;
>     }
> 
>     public bool CanGoBack => this.frame.CanGoBack;
> 
>     public void GoBack() => this.frame.GoBack();
> 
>     public void Navigate<T>(object args)
>     {
>         this.frame.Navigate(this.viewMapping[typeof(T)], args);
>     }
> }
> ```
> 
> So that then you could use the navigation service from a framework-agnostic viewmodel, by just doing:
> 
> ```cs
> this.navigationService.Navigate<SomeViewModel>();
> ```
> 
> And the platform specific service would look up the target page type automatically.
> 
> This is just the general idea, as I've mentioned there could be plenty of other customizations for this, like:
> 
> * Having APIs to add/update the mapped types after the instantiation of the service
> * Having some APIs to control other navigation parameters, reset the stack, etc.
> * Custom navigation types, eg. across multiple navigation frames (I do this in my apps, for instance), etc.
> 
> We could definitely add a sample docs page for this in the future though 😄

Good day!

Are you sure this will work correctly? After all, you need to insert either a Uri or a Page Instance into the Navigate - otherwise it will output the full type name to the view (override ToString)
![image](https://github.com/CommunityToolkit/MVVM-Samples/assets/88714829/5f3fa278-786f-4fda-b5b2-7884538c5b4e)
```XAML
        <Frame Grid.Row="1" NavigationUIVisibility="Visible" x:Name="MainFrame"/>
```

**Comment by GF-Huang** [comment]
But all the above examples losing control of `NavigationTransitionInfo` for animating.
